x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java

```diff
@@ -555,6 +563,9 @@ private void prewarmCache(ActionListener<Void> listener) {
                 }
             } catch (IOException e) {
                 logger.warn(() -> new ParameterizedMessage("{} unable to prewarm file [{}]", shardId, file.physicalName()), e);
+                if (submitted == false) {
+                    completionListener.onFailure(e);
```

Comment:
I think this attempts to ensure that the recovery state moves to `DONE`? I guess that is more correct than `FINALIZE`, but I think we should add a test for it.

Reply:
It attempts to complete the grouped listener but has not other side effect besides logging a warning message if the shard is a relocation target: https://github.com/elastic/elasticsearch/blob/a5d2251064d3457dc253503ba9ce75dd57fa6f4e/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotIndexEventListener.java#L78-L90

I'm OK with leaving the recovery state in `FINALIZE` as this is what allowed to detect a non-fully prewarmed shard at the first place. I think that adding more logs would have helped to troubleshoot some issues when accessing repository files.